polkadot/node/core/pvf/src/execute/queue.rs

```diff
@@ -608,3 +622,181 @@ pub fn start(
+	counter: HashMap<PvfExecPriority, usize>,
+}
+
+impl Unscheduled {
+	// A threshold reaching which we reset counted jobs.
+	// Max number of jobs per block assuming 6s window, 2 CPU cores, and 2s for a run.
+	const MAX_COUNT: usize = 12;
+	// A threshold in percentages, the portion a current priority can "steal" from lower ones.
+	// For example:
+	// Disputes take 70%, leaving 30% for approvals and all backings.
+	// 80% of the remaining goes to approvals, which is 30% * 80% = 24% of the original 100%.
+	// If we used parts of the original 100%, approvals can't take more than 24%,
+	// even if there are no disputes.
+	const FULFILLED_THRESHOLDS: &'static [(PvfExecPriority, usize)] = &[
+		(PvfExecPriority::Dispute, 70),
+		(PvfExecPriority::Approval, 80),
+		(PvfExecPriority::BackingSystemParas, 100),
+		(PvfExecPriority::Backing, 100),
+	];
+
+	fn new() -> Self {
+		Self {
+			unscheduled: PvfExecPriority::iter()
+				.map(|priority| (priority, VecDeque::new()))
+				.collect(),
+			counter: PvfExecPriority::iter().map(|priority| (priority, 0)).collect(),
+		}
+	}
+
+	fn select_next_priority(&self) -> PvfExecPriority {
+		PvfExecPriority::iter()
+			.find(|priority| self.has_pending(priority) && !self.is_fulfilled(priority))
+			.unwrap_or_else(|| {
+				PvfExecPriority::iter()
+					.find(|priority| self.has_pending(priority))
+					.unwrap_or(PvfExecPriority::Backing)
+			})
+	}
+
+	fn get_mut(&mut self, priority: PvfExecPriority) -> Option<&mut VecDeque<ExecuteJob>> {
+		self.unscheduled.get_mut(&priority)
+	}
+
+	fn add(&mut self, job: ExecuteJob, priority: PvfExecPriority) {
+		self.unscheduled.entry(priority).or_default().push_back(job);
+	}
+
+	fn has_pending(&self, priority: &PvfExecPriority) -> bool {
+		!self.unscheduled.get(priority).unwrap_or(&VecDeque::new()).is_empty()
+	}
+
+	fn fulfilled_threshold(priority: &PvfExecPriority) -> Option<usize> {
+		Self::FULFILLED_THRESHOLDS.iter().find_map(
+			|&(p, value)| {
+				if p == *priority {
+					Some(value)
+				} else {
+					None
+				}
+			},
+		)
+	}
+
+	fn is_fulfilled(&self, priority: &PvfExecPriority) -> bool {
+		let Some(threshold) = Self::fulfilled_threshold(priority) else { return false };
+		let Some(count) = self.counter.get(&priority) else { return false };
+		// Every time we iterate by lower level priorities
+		let total_count: usize = self
+			.counter
+			.iter()
+			.filter_map(|(p, c)| if *p >= *priority { Some(c) } else { None })
+			.sum();
+		if total_count == 0 {
+			return false
+		}
+
+		// Because we operate through a small range, we can't let a priority go over the
+		// threshold, so we check fulfillment by adding one more run
+		(count + 1) * 100 / total_count >= threshold
+	}
+
+	fn log(&mut self, priority: PvfExecPriority) {
+		let current_count: &mut usize = self.counter.entry(priority).or_default();
+		*current_count += 1;
+
+		if self.counter.values().sum::<usize>() >= Self::MAX_COUNT {
+			self.reset_counter();
```

Comment:
I think we discussed that if the queue is not above MAX_COUNT we would want to prioritize backing or at least backing for system chains, because backing is a bit more time sensitive than approval, the new refactor doesn't seem to cover that.

Are we ok with this trade-off ? If yes, can we find a way to properly log when this condition is met, so that we can at least detect it if it happens in production.

Reply:
With the current logic with max loaded queues we should run a backing job once per 12 execution. Backing system parachains has more priority than just backing.